### ISSUE TITLE
Broadcast fusion fails when scalars are involved

### ISSUE THREAD

**Issue by ChrisRackauckas** [JuliaGPU/GPUArrays.jl#22]
```julia
u0 = GPUArray(rand(Float32, 32, 32))
tmp = ones(u0)
uprev = ones(u0)
k1 = ones(u0)
a = Float32(2.0)
tmp .=  uprev.+a.*k1
```

```
indexing not defined for GPUArrays.GPUArray{Float32,2,OpenCL.cl.Buffer{Float32},GPUArrays.CLBackend.CLContext}
getindex at abstractarray.jl:874 [inlined]
_getindex at abstractarray.jl:921 [inlined]
getindex at abstractarray.jl:875 [inlined]
_broadcast_getindex at broadcast.jl:133 [inlined]
_broadcast_getindex at broadcast.jl:130 [inlined]
macro expansion at broadcast.jl:151 [inlined]
macro expansion at simdloop.jl:73 [inlined]
macro expansion at broadcast.jl:147 [inlined]
_broadcast!(::##5#6, ::GPUArrays.GPUArray{Float32,2,OpenCL.cl.Buffer{Float32},GPUArrays.CLBackend.CLContext}, ::Tuple{Tuple{Bool,Bool},Tuple{},Tuple{Bool,Bool}}, ::Tuple{Tuple{Int64,Int64},Tuple{},Tuple{Int64,Int64}}, ::GPUArrays.GPUArray{Float32,2,OpenCL.cl.Buffer{Float32},GPUArrays.CLBackend.CLContext}, ::Tuple{Float32,GPUArrays.GPUArray{Float32,2,OpenCL.cl.Buffer{Float32},GPUArrays.CLBackend.CLContext}}, ::Type{Val{2}}, ::CartesianRange{CartesianIndex{2}}) at broadcast.jl:139
broadcast_c! at broadcast.jl:211 [inlined]
broadcast!(::Function, ::GPUArrays.GPUArray{Float32,2,OpenCL.cl.Buffer{Float32},GPUArrays.CLBackend.CLContext}, ::GPUArrays.GPUArray{Float32,2,OpenCL.cl.Buffer{Float32},GPUArrays.CLBackend.CLContext}, ::Float32, ::GPUArrays.GPUArray{Float32,2,OpenCL.cl.Buffer{Float32},GPUArrays.CLBackend.CLContext}) at broadcast.jl:204
include_string(::String, ::String) at loading.jl:515
include_string(::String, ::String, ::Int64) at eval.jl:30
include_string(::Module, ::String, ::String, ::Int64, ::Vararg{Int64,N} where N) at eval.jl:34
(::Atom.##49#52{String,Int64,String})() at eval.jl:50
withpath(::Atom.##49#52{String,Int64,String}, ::String) at utils.jl:30
withpath(::Function, ::String) at eval.jl:38
macro expansion at eval.jl:49 [inlined]
(::Atom.##48#51{Dict{String,Any}})() at task.jl:80
```

**Comment by SimonDanisch** [comment]
That probably just needs another broadcast overload .... It's a bit problematic to overload broadcast without ambiguities. I should try to have a more systematic approach of what I overload!

**Comment by ChrisRackauckas** [comment]
I actually don't understand how you're defining broadcast here. In my packages I had to make sure that broadcasts could include `Number` and used a generating function to build the correct call. I would assume something similar would work here.

**Comment by SimonDanisch** [comment]
I wanted it to work with `Any` as well, which makes it more ambiguous, I think...

**Comment by ChrisRackauckas** [comment]
>I wanted it to work with Any as well, which makes it more ambiguous, I think...

Not sure how that could work since you need to be making a GPU kernel, right? At least I assumed what you're somehow doing is building an anonymous function and generating a GPU kernel to map on that. And if that's what you're doing, handling scalars is fine, but arbitrary Julia types will definitely not work.

**Comment by SimonDanisch** [comment]
immutable arbitrary julia type work relatively fine ;)

**Comment by SimonDanisch** [comment]
Solved by #37
